runtime-sdk/src/modules/core/mod.rs

```diff
@@ -51,3 +119,135 @@ pub mod state {
     /// Map of message idx to message handlers for messages emitted in previous round.
     pub const MESSAGE_HANDLERS: &[u8] = &[0x02];
 }
+
+pub struct Module;
+
+const CONTEXT_KEY_GAS_USED: &str = "core.GasUsed";
+
+impl Module {
+    /// Initialize state from genesis.
+    fn init(ctx: &mut DispatchContext<'_>, genesis: &Genesis) {
+        // Set genesis parameters.
+        Self::set_params(ctx.runtime_state(), &genesis.parameters);
+    }
+
+    /// Migrate state from a previous version.
+    fn migrate(_ctx: &mut DispatchContext<'_>, _from: u32) -> bool {
+        // No migrations currently supported.
+        false
+    }
+}
+
+impl API for Module {
+    fn use_gas<C: Context>(ctx: &mut C, gas: u64) -> Result<(), Error> {
+        let new_gas_used = match (
+            // todo: oh shoot we can't borrow for both of these at the same time
```

Comment:
Yeah you'll need to refactor this. Probably just do something like `ctx.tx_auth_info().map(|ai| ai.fee.gas)` to avoid holding a reference to `ctx` just for the gas limit specified in the transaction.

Reply:
yup